expression/builtin_arithmetic.go

```diff
@@ -0,0 +1,270 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expression
+
+import (
+	"math"
+
+	"fmt"
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/parser/opcode"
+	"github.com/pingcap/tidb/util/types"
+)
+
+var (
+	_ functionClass = &arithmeticPlusFunctionClass{}
+	_ functionClass = &arithmeticFunctionClass{}
+)
+
+var (
+	_ builtinFunc = &builtinArithmeticPlusRealSig{}
+	_ builtinFunc = &builtinArithmeticPlusIntSig{}
+	_ builtinFunc = &builtinArithmeticPlusIntUnsignedSig{}
+	_ builtinFunc = &builtinArithmeticPlusDecimalSig{}
+	_ builtinFunc = &builtinArithmeticSig{}
+)
+
+func numericContextResultType(ft *types.FieldType) types.TypeClass {
+	if types.IsTemporalType(ft.Tp) {
+		if ft.Decimal > 0 {
+			return types.ClassDecimal
+		}
+		return types.ClassInt
+	}
+	if ft.ToClass() == types.ClassString {
+		return types.ClassReal
+	}
+	return ft.ToClass()
+}
+
+type arithmeticPlusFunctionClass struct {
+	baseFunctionClass
+}
+
+// setFlenDecimal4Int is called to set proper `Flen` and `Decimal` of return
+// type according to the two input parameter's types.
+func (c *arithmeticPlusFunctionClass) setFlenDecimal4Int(retTp, a, b *types.FieldType) {
+	retTp.Decimal = 0
+	if a.Flen == types.UnspecifiedLength || b.Flen == types.UnspecifiedLength {
+		retTp.Flen = mysql.MaxIntWidth
+	} else {
+		retTp.Flen = int(math.Max(float64(a.Flen), float64(b.Flen))) + 1
```

Comment:
I think we can set retTp to mysql.MaxIntWidth,
the length of int is not that important,
set it to the default value make the code more readable.